File: Chapter 2/Section 2.3/nocows.cpp
/*
ID: hsfncd31
TASK: nocows
LANG: C++                 
*/
#include <fstream>

static const int MAXN = 200 - 1, MAXK = 100 - 1, MOD = 9901;
int dp[MAXN + 1][MAXK + 1];

int main() {
    std::fstream in("nocows.in", std::fstream::in), out("nocows.out", std::fstream::out);

    int N, K;

    in >> N >> K;
    dp[0][0] = dp[1][1] = 1;
    for (int n = 2; n <= N; ++n)
        for (int k = 0; k <= std::min(n, K); ++k)
            for (int left_n = 1; left_n < n - 1; ++left_n) {
                for (int smaller_k = 1; smaller_k < k - 1; ++smaller_k) {
                    dp[n][k] += dp[left_n][k - 1] * dp[n - 1 - left_n][smaller_k];
                    dp[n][k] += dp[left_n][smaller_k] * dp[n - 1 - left_n][k - 1];
                    dp[n][k] %= MOD;
                }
                dp[n][k] += dp[left_n][k - 1] * dp[n - 1 - left_n][k - 1];
                dp[n][k] %= MOD;
            }
    out << dp[N][K] << '\n';

    in.close();
    out.close();
}

File: Chapter 2/Section 2.3/nocows.py
"""
ID: hsfncd31
LANG: PYTHON3
TASK: nocows
Time Limit Exceeded in Case#7
"""
import os
import typing


def main():
    base_filename = 'test' if os.name == 'nt' else 'nocows'
    with open('{}.in'.format(base_filename), 'r') as infile,\
            open('{}.out'.format(base_filename), 'w') as outfile:  # type: typing.IO[str]
        def get_line() -> str:
            return infile.readline().rstrip('\n')

        def out_print(*args, **kwargs) -> None:
            kwargs['file'] = outfile
            print(*args, **kwargs)

        N, K = map(int, get_line().split())
        max_n, max_k, mod = 200 - 1, 100 - 1, 9901
        # n, k
        dp = [[0] * (max_k + 1) for _ in range(max_n + 1)]
        dp[0][0], dp[1][1] = 1, 1
        for n in range(2, N + 1):
            for k in range(min(len(dp[0]), n + 1, K + 1)):
                for left_n in range(1, n - 1):
                    for smaller_k in range(k - 1):
                        dp[n][k] += dp[left_n][k - 1] * dp[n - 1 - left_n][smaller_k]
                        dp[n][k] += dp[left_n][smaller_k] * dp[n - 1 - left_n][k - 1]
                    # both k - 1
                    dp[n][k] += dp[left_n][k - 1] * dp[n - 1 - left_n][k - 1]
                    dp[n][k] %= mod
                # print(n, k, dp[n][k])
        out_print(dp[N][K])


if __name__ == '__main__':
    main()
